Builtins

porglezomp opened the pull request:
Add builtins!

This is WIP pending adding tests for these.
This also changes the codegen to not require static binding for global variables, and instead have lookup fail at runtime.

porglezomp committed:
Add instruction intrinsics

porglezomp committed:
Don't throw an error if a symbol is not statically available

porglezomp committed:
Implement the builtin functions

porglezomp committed:
Add the prelude

james-perretta commented on runner.py:
```diff
@@ -0,0 +1,173 @@
+def inst_function(
+        name: SSym, params: List[Var], *insts: Inst,
+        should_return: bool = True) -> SFunction:
+    """Create a function that's just one instruction."""
```
Shouldn't this say "...consisting of the instructions in `insts`"?

porglezomp replied:
Ope, yep, changed that as I needed to refactor `inst/store`.

james-perretta commented on runner.py:
```diff
@@ -0,0 +1,173 @@
+    for inst in insts:
+        begin.add_inst(inst)
+    if should_return:
+        begin.add_inst(bytecode.ReturnInst(Var('result')))
```
`Var('result')` is a bit "magic string"-y for my tastes. What if this function took in a variable to return instead of a bool (or None if there's no return value)?

james-perretta commented on runner.py:
```diff
@@ -0,0 +1,173 @@
+
+def add_builtins(env: Dict[SSym, Value]) -> None:
+    """Add builtins to the environment."""
+    add_intrinsics(env)  # @TODO: Don't do this, for greater flexibility?
```
I agree, we should just call `add_builtins`, `add_intrinsics`, and `add_prelude` in our setup code.

james-perretta commented on runner.py:
```diff
@@ -0,0 +1,173 @@
+    add_intrinsics(env)  # @TODO: Don't do this, for greater flexibility?
+    code = scheme.parse("""
+    (define (trap) (inst/trap))
+    (define (assert b) (if b 0 (trap)))
```
Good call on adding `assert`

james-perretta commented on runner.py:
```diff
@@ -0,0 +1,173 @@
+      (assert (number< n (vector-length v)))
+      (inst/store v n x))
+
+    (define (vector-make/recur len idx v x)
```
Could you add a a comment/example of what this does? (same for `vector=/recur` later on)

porglezomp replied:
It's just needed to build the loop inside the function, since they need parameters than we expose. It's not intended to be used (in a more capable language I would've had it nested inside).

porglezomp replied:
Also… what do you mean by this? I can add comments in the source code if we merge #37 first.

james-perretta replied:
Ah ok, I wasn't sure if it was an extra builtin function or just an implementation detail.

james-perretta commented on runner.py:
```diff
@@ -0,0 +1,173 @@
+
+
+def add_prelude(env: Dict[SSym, Value]) -> None:
+    """Add intrinsics to the environment."""
```
Copy-pasta?

porglezomp committed:
Refactor inst_function to take the return var name

porglezomp committed:
Move the type predicates into builtins

porglezomp committed:
Add builtin arithmetic operators

porglezomp committed:
Add functions to run code

porglezomp committed:
Small instruction changes

Fix inst/%
Move not into builtins
Make / and % assert that the second argument is nonzero

porglezomp committed:
Minor bugfixes

Correctly parse negative number literals
Fix assertion in CallInst

porglezomp committed:
Add tests for intrinsics

porglezomp committed:
Add breakpoint and trace instructions and builtins

Resolves #38

porglezomp committed:
Prevent call instructions from running twice

!!!!!!!!!

porglezomp committed:
Add tests for builtins

porglezomp committed:
Add some tests for prelude

james-perretta commented on runner.py:
```diff
@@ -0,0 +1,251 @@
+    """Add intrinsics to the environment."""
+    def inst_function(
+            name: SSym, params: List[Var],
+            return_to: Optional[bytecode.Parameter], *insts: Inst,
```
Suggest renaming since Parameters aren't necessarily storage

james-perretta commented on runner.py:
```diff
@@ -0,0 +1,251 @@
+        emitter.visit(definition)
+
+
+name_counter = 0
```
Is this only used in one place?

james-perretta commented on test_builtins.py:
```diff
@@ -0,0 +1,130 @@
+            run(env, '(symbol= 1 0)')
+        self.assertEqual(run(env, "(symbol= 'a 'b)"), SBool(False))
+        self.assertEqual(run(env, "(symbol= 'a 'a)"), SBool(True))
+        # I don't want to test the other foo= functions, they're very similar
```
I got you a present:
```
        # This first one has an uncaught exception because there aren't any type checks for pointer=. You can decide what to do about that.
        with self.assertRaises(errors.Trap, msg="(trap)"):
            run(env, "(pointer= 'a false)")
        self.assertEqual(run(env, "(pointer= (lambda () 0) (lambda () 0))"),
                         SBool(False))
        self.assertEqual(run(env, "(define (func) 42) (pointer= func func)"),
                         SBool(True))
        self.assertEqual(run(env, "(pointer= 0 0)"), SBool(True))

        with self.assertRaises(errors.Trap, msg="(trap)"):
            run(env, "(number= 'num 1)")
        self.assertEqual(run(env, "(number= 42 43)"), SBool(False))
        self.assertEqual(run(env, "(number= 42 42)"), SBool(True))

        with self.assertRaises(errors.Trap, msg="(trap)"):
            run(env, "(number< 'num 1)")
        self.assertEqual(run(env, "(number< 42 42)"), SBool(False))
        self.assertEqual(run(env, "(number< 42 43)"), SBool(True))
```

porglezomp committed:
Review fixes
